### PR TITLE
8340831: Simplify simple validation for class definition in MethodHandles.Lookup

### PR DESCRIPTION
**Pull request by liach**
`MethodHandles.Lookup` defines a `ClassFile` for simple validations; it is unnecessary and can be scalarized manually. The removal of `ClassFile` class is also slightly helpful for bootstrap by reducing class loading. Also improved class file version checking in `VM` class.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340831](https://bugs.openjdk.org/browse/JDK-8340831): Simplify simple validation for class definition in MethodHandles.Lookup (**Sub-task** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21170/head:pull/21170` \
`$ git checkout pull/21170`

Update a local copy of the PR: \
`$ git checkout pull/21170` \
`$ git pull https://git.openjdk.org/jdk.git pull/21170/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21170`

View PR using the GUI difftool: \
`$ git pr show -t 21170`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21170.diff">https://git.openjdk.org/jdk/pull/21170.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21170#issuecomment-2372479334)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -2242,85 +2244,70 @@ private static ClassFileDumper defaultDumper() {
...
-                }
-                return cf;
+            String name;
+            ClassDesc sym;
+            int accessFlags;
+            try {
+                ClassModel cm = ClassFile.of().parse(bytes);
+                var thisClass = cm.thisClass();
+                name = thisClass.asInternalName();
+                sym = thisClass.asSymbol();
```
Comment (cl4es) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2276]: We only use this for determining package name are equal, and sym.packageName() does the similar transformations plus a bit more. Likely not a significant cost compared to the `ClassFile::parse` - but perhaps there's room for a utility method to get the package name directly from an internal name?

**Review by cl4es — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2276]: This is not part of internal code path so I don't think it is that sensitive

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -2446,23 +2434,12 @@ private ClassDefiner makeHiddenClassDefiner(ClassFile cf,
...
-            private final ClassFileDumper dumper;
-
-            private ClassDefiner(Lookup lookup, ClassFile cf, int flags, ClassFileDumper dumper) {
-                assert ((flags & HIDDEN_CLASS) != 0 || (flags & STRONG_LOADER_LINK) == STRONG_LOADER_LINK);
-                this.lookup = lookup;
-                this.bytes = cf.bytes;
-                this.name = cf.name;
-                this.classFlags = flags;
-                this.dumper = dumper;
+        record ClassDefiner(Lookup lookup, String name, byte[] bytes, int classFlags, ClassFileDumper dumper) {
```
[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2440]: Should I rename other existing name vars too?

[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2440]: AFAICT they already use the record names, so we're not generating redundant code there.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (cl4es) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2276]: Agree - was just thinking out loud on this one. What you have is good enough here.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2440]: I mean the name variable in makeHiddenClassDefiner and makeClassDefiner that leads up here.

[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2440]: Ah.. Yeah, either way you like, but if it's an internal name (`foo/Bar`) then including internal in the variable names consistently is good I think.

**Review by cl4es — APPROVED** [review]

**Comment by cl4es** [comment]
> Apparently forgot to push when I re-requested a review after the renames.

I thought something was off, but on the third refresh the changes came in :-)

**Comment by mlchung** [comment]
This change looks okay (belated review).

**Pull request closed without merging**